include/ignition/gui/qml/GzPose.qml

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+import QtQuick 2.9
+import QtQuick.Controls 1.4
+import QtQuick.Controls 2.2
+import QtQuick.Controls.Material 2.1
+import QtQuick.Layouts 1.3
+import QtQuick.Controls.Styles 1.4
+
+// Item displaying 3D pose information.
+Item {
+  id: gzPoseRoot
+  height: gzPoseContent.height
+
+  // Left indentation
+  property int indentation: 10
+
+  // Horizontal margins
+  property int margin: 5
+
+  // Maximum spinbox value
+  property double spinMax: 1000000
+
+  // Read-only / write
+  property bool readOnly: false
+
+  // Show Pose bar
+  property bool show: false
+
+  // Loaded item for Pose
+  property var xItem: {}
+  property var yItem: {}
+  property var zItem: {}
+  property var rollItem: {}
+  property var pitchItem: {}
+  property var yawItem: {}
+
+  property double xModelValue
+  property double yModelValue
+  property double zModelValue
+  property double rollModelValue
+  property double pitchModelValue
+  property double yawModelValue
```

Comment:
Using one set is possible but that would make using the common widget extra tedious. We need to write this piece of code 6 times when we declare GzPose: 
```
text: {
  var decimals = getDecimals(numberText.width)
  return model[0].toFixed(decimals)
}
```

Reply:
I agree that it would be nice to use the same set of variables to read and write, otherwise the user of `GzPose` needs to know that they set `xModelValue`, but need to read from `xItem.value`, which can become confusing.

I believe what we want here is to expose the same variable to the user of `GzPose` that they can read and write from. I don't have strong feelings about whether the pose needs to be combined into a single variable, or exposed as 6 separate ones. I do think we should drop the `Model` from the variable names, and just expose `xValue`, `yValue`, etc.

I think the piece we're missing here is that the loader is setting `numberValue` from `xModelValue`, but it never updates `xModelValue` when the value changes. In other words, we have 1-way-binding, but we'd like 2-way-binding.

This extra binding rule worked for me, so that whenever the value of the spin changes, `xModelValue` is also updated.

```diff
diff --git a/include/ignition/gui/qml/GzPose.qml b/include/ignition/gui/qml/GzPose.qml
index 72bfb08c..e1931aa1 100644
--- a/include/ignition/gui/qml/GzPose.qml
+++ b/include/ignition/gui/qml/GzPose.qml
@@ -131,6 +131,11 @@ Item {
           id: xLoader
           anchors.fill: parent
           property double numberValue: xModelValue
+          Binding {
+            target: gzPoseRoot
+            property: "xModelValue"
+            value: xLoader.item.value
+          }
           sourceComponent: readOnly ? readOnlyNumber : writableNumber
           onLoaded: {
             xItem = xLoader.item
```

This way, we can remove all the `*Item` variables, and directly call `gzPoseContent.xModelValue` on `gz-sim` instead of `gzPoseContent.xItem.value`.

Reply:
I think this will cause another bug: (I only use `Binding` for X, Y is using my old code) The X value in the spinbox for the cylinder is overwritten by the X value of the sphere.
![common_pose_bug](https://user-images.githubusercontent.com/50132891/175427677-ea1b1a3c-e3e5-49b9-b5fa-134cc6e6034a.gif)

Reply:
Ahh ok, so the cyclical 2-way binding was probably not a good idea. How about then sending the spin values with the `gzPoseSet` signal itself? This way the receiver doesn't need to access those values directly, as they will come with the callback.

Reply:
I think we still need 2 sets of parameters in `GzPose.qml` even with this method. Here's why:

Since all `signal gzPoseSet()`'s are called from the same `Component`, there is no way for this component to know which value (x/y/z/r/p/y) is changed without adding additional signals. So we need to keep 6 local variables that keep track of all the values of the spinboxes.

Even worse, since `id` in QML is only accessible during compile time (that's said, `property double a: Loader.item.SpinBox.value` is illegal, because SpinBox is `id`). We still need to keep the same `property var xItem: {}`'s. Now, the users can finally read from the signal parameters, but they can also directly read from `xItem.value`. 

Conclusion: Yes, this change will make usage of this common widget less confusing, but it also introduces redundant parameters (signal parameters VS properties).

Reply:
>  there is no way for this component to know which value (x/y/z/r/p/y) is changed without adding additional signals

How about `gzPoseSet` always sending all the current elements of a pose? I haven't tried it, but I'm thinking of something like this:

```diff
diff --git a/include/ignition/gui/qml/GzPose.qml b/include/ignition/gui/qml/GzPose.qml
index 72bfb08c..dc9675f0 100644
--- a/include/ignition/gui/qml/GzPose.qml
+++ b/include/ignition/gui/qml/GzPose.qml
@@ -56,7 +56,7 @@ Item {
   property double pitchModelValue
   property double yawModelValue
 
-  signal gzPoseSet()
+  signal gzPoseSet(x: double, y: double, y: double, roll: double, pitch: double, yaw: double)
 
   /**
    * Used to create a spin box
@@ -70,7 +70,7 @@ Item {
       maximumValue: spinMax
       decimals: getDecimals(writableSpin.width)
       onEditingFinished: {
-        gzPoseRoot.gzPoseSet()
+        gzPoseRoot.gzPoseSet(xItem.value, yItem.value, zItem.value...)
       }
     }
   }
```

> We still need to keep the same property var xItem: {}'s

Yeah we'll need to keep them, but that's going to be an internal implementation detail that users of `GzPose` don't need to worry about. Sure, in Javascript we can't really make anything private, but we shouldn't encourage users to directly access the internals of the component.

Let me know what you think!

Reply:
LGTM. I'll try to implement this.